newrelic/newrelic-browser-agent
# add build process and test for full e2e microfrontend experience

### Overview 

This PR adds a lot of testing functionality to the repo.  Please refer to the following functionalities added.

1. Adds a true build of a micro front end, which uses `npm pack` on our NPM package to truly test it's integration.
    - JIL can run tests on an actual micro front end
2. Extended JIL promises to accept an `appID` argument
    - This allows payload promises to resolve only when they receive a payload from the correct appID
    - This is useful when there is more than one agent on the page, something that has never been tested before now
3. Added recursive `builder.js` file, which will allow us to add more `test-builds` in the future to test against, if warranted.

### Manual testing

Steps to validate:
1. run `npm install`
2. run `npm run tools:test-builds`
    - This will build the NPM package and the test-builds (the microfrontend sandboxes that JIL can run against)
    - This is a command that could be used in GH actions to get all the builds created before testing.
    - This will also generate a `.tgz` file from our NPM package, and allow our sandbox envs to install it, to test the real world ability to install our packages.
4. Run `npm run test -- -v -f merged --retry false -b chrome@latest tests/functional/modules/js-errors/js-errors-multiple-scopes.test.js` to see an example of a JIL test running against an actual microfrontend build.
